File: fsharp/ledger/Ledger.fs
module Ledger

open System
open System.Globalization

type Entry = { date: DateTime; description: string; change: int }

let mkEntry date description change = { 
        date = DateTime.Parse(date, CultureInfo.InvariantCulture); 
        description = description; change = change 
    }
        
let formatLedger currency (locale:string) entries = 
    let culture = new CultureInfo(locale)
    culture.NumberFormat.CurrencySymbol <- match currency with | "EUR" -> "€" | _ -> "$"
    let dateFormat = match locale with | "nl-NL" -> "dd-MM-yyyy" | _ -> "MM\/dd\/yyyy"

    let linePrinter entry = 
        let date = entry.date.ToString(dateFormat)     
        let description = 
            match entry.description.Length with
            | l when l > 25 ->  entry.description.[0..21] + "..."
            | _ -> entry.description
        let change = String.Format(culture, "{0:C}{1}", float entry.change / 100.0, 
                        match entry.change with | _ when entry.change < 0 -> "" | _ -> " ")

        sprintf "%s | %-25s | %13s" date description change

    System.String.Join("\n", seq {
        match locale with
        | "nl-NL" ->    yield "Datum      | Omschrijving              | Verandering  "
        | _ ->          yield "Date       | Description               | Change       "

        let sorted = List.sortBy (fun x -> x.date, x.description, x.change) entries
        yield! sorted |> List.map linePrinter
    })

File: csharp/ledger/Ledger.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

public class LedgerEntry
{
    public DateTime Date { get; }
    public string Desc { get; }
    public float Chg { get; }

    public LedgerEntry(DateTime date, string desc, float chg) => 
        (Date, Desc, Chg) = (date, desc, chg);
}

public static class Ledger
{
    public static LedgerEntry CreateEntry(string date, string desc, int chng) =>
        new LedgerEntry(DateTime.Parse(date, CultureInfo.InvariantCulture), desc, chng / 100.0f);

    private static CultureInfo CreateCulture(string cur, string loc)
    {
        var culture = new CultureInfo(loc);
        culture.NumberFormat.CurrencySymbol = cur == "USD" ? "$" : "€";
        culture.NumberFormat.CurrencyNegativePattern = loc == "en-US" ? 0 : 12;
        culture.DateTimeFormat.ShortDatePattern = loc == "en-US" ? "MM/dd/yyyy" : "dd/MM/yyyy";
        return culture;
    }

    private static string PrintHead(string loc) =>
        loc == "en-US"
            ? "Date       | Description               | Change       "
            : "Datum      | Omschrijving              | Verandering  ";

    private static string PrintEntry(LedgerEntry entry)
    {
        var description = entry.Desc.Length > 25 ? entry.Desc.Substring(0, 22) + "..." : entry.Desc;
        var space = entry.Chg < 0.0 ? "" : " ";
        return $"{entry.Date:d} | {description,-25} | {entry.Chg,13:C}{space}";
    }

    private static IEnumerable<LedgerEntry> Sort(LedgerEntry[] entries) =>
        entries.Where(e => e.Chg < 0).OrderBy(x => $"{x.Date}@{x.Desc}@{x.Chg}")
        .Concat(entries.Where(e => e.Chg >= 0).OrderBy(x => $"{x.Date}@{x.Desc}@{x.Chg}"));

    public static string Format(string currency, string locale, LedgerEntry[] entries)
    {
        CultureInfo.DefaultThreadCurrentCulture = CreateCulture(currency, locale);
        return string.Join('\n', new[] { PrintHead(locale) }.Concat(Sort(entries).Select(PrintEntry)));
    }
}
